pkg/agent/reconcile_suspended_pod_termination.go

```diff
@@ -0,0 +1,91 @@
+package agent
+
+import (
+	"context"
+	"fmt"
+	"os/exec"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/events"
+	"k8s.io/klog/v2"
+
+	"github.com/piraeusdatastore/piraeus-ha-controller/pkg/metadata"
+)
+
+type suspendedPodReconciler struct {
+	opt *Options
+}
+
+var _ Reconciler = &suspendedPodReconciler{}
+
+// NewSuspendedPodReconciler creates a reconciler that gets suspended Pods to resume termination.
+//
+// While DRBD is suspending IO, all processes (including Pods and filesystems) using the device are stuck. In order
+// to resume, one can force DRBD to report IO errors instead. The reconciler does just that if a local Pod should
+// be stopped while it is suspended by DRBD. This enables a (relatively) clean shutdown of the resource without
+// node reboot.
+func NewSuspendedPodReconciler(opt *Options) Reconciler {
+	return &suspendedPodReconciler{
+		opt: opt,
+	}
+}
+
+func (s *suspendedPodReconciler) RunForResource(ctx context.Context, req *ReconcileRequest, recorder events.EventRecorder) error {
+	klog.V(3).Infof("checking if resource needs to be forced to become secondary")
+
+	if !req.Resource.Suspended {
+		klog.V(4).Infof("resource '%s' not suspended", req.Resource.Name)
+		return nil
+	}
+
+	if !req.Resource.Primary() {
+		klog.V(4).Infof("resource '%s' not suspended", req.Resource.Name)
+		return nil
+	}
+
+	// NB: we use runtime.Object here, so we don't accidentally create a "nil interface value", which would
+	// confuse the event-recorder into panicking.
+	var matchingVA runtime.Object
+	for _, va := range req.Attachments {
+		if va.Spec.NodeName == s.opt.NodeName {
+			matchingVA = va
+			break
+		}
+	}
+
+	for _, pod := range req.Pods {
+		if pod.Spec.NodeName != s.opt.NodeName {
+			klog.V(4).Infof("Pod '%s/%s' not running on this node", pod.Namespace, pod.Name)
+			continue
+		}
+
+		if pod.Status.Phase == corev1.PodRunning && pod.DeletionTimestamp == nil {
+			klog.V(4).Infof("consumer pod '%s/%s' of resource '%s' is not terminating, no forced demotion", pod.Namespace, pod.Name, req.Resource.Name)
+			return nil
+		}
+	}
+
+	klog.V(1).Infof("force demotion of resource '%s'", req.Resource.Name)
+
+	cmd := []string{"drbdsetup", "secondary", "--force", req.Resource.Name}
```

Comment:
you know that you need to run `secondary --force` at that point in time? otherwise - in general - something that calls `secondary` and then `secondary --force` might be nicer? We have that in the shim in drbd-utils now (unreleased).

Reply:
Added a comment explaining the situation. basically the resource is suspended and primary, and since we use auto-promote that means only `secondary --force` can work.